cpp/src/arrow/compute/kernels/sort_to_indices.cc

```diff
@@ -107,37 +176,34 @@ class SortToIndicesKernelImpl : public SortToIndicesKernel {
     int64_t* indices_begin = reinterpret_cast<int64_t*>(indices_buf->mutable_data());
     int64_t* indices_end = indices_begin + values->length();
 
-    std::iota(indices_begin, indices_end, 0);
-    auto nulls_begin = indices_end;
-    if (values->null_count()) {
-      nulls_begin =
-          std::stable_partition(indices_begin, indices_end,
-                                [&values](uint64_t ind) { return !values->IsNull(ind); });
-    }
-    std::stable_sort(indices_begin, nulls_begin,
-                     [&values, this](uint64_t left, uint64_t right) {
-                       return compare_(*values, left, right);
-                     });
+    sorter_.Sort(indices_begin, indices_end, *values.get());
     *offsets = std::make_shared<UInt64Array>(values->length(), indices_buf);
     return Status::OK();
   }
 };
 
-template <typename ArrowType, typename Comparator>
-SortToIndicesKernelImpl<ArrowType, Comparator>* MakeSortToIndicesKernelImpl(
+template <typename ArrowType, typename Comparator,
+          typename Sorter = CompareSorter<ArrowType, Comparator>>
+SortToIndicesKernelImpl<ArrowType, Sorter>* MakeSortToIndicesKernelImpl(
     Comparator comparator) {
-  return new SortToIndicesKernelImpl<ArrowType, Comparator>(comparator);
+  return new SortToIndicesKernelImpl<ArrowType, Sorter>(Sorter(comparator));
+}
+
+template <typename ArrowType, typename Sorter = CountSorter<ArrowType>>
+SortToIndicesKernelImpl<ArrowType, Sorter>* MakeSortToIndicesKernelImpl(int min,
```

Comment:
For clarity, perhaps give different names to those two overloads? For example `MakeSortToIndicesWithComparator` and `MakeSortToIndicesCounting`?

Reply:
done